File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def getLength(self, head):
        count = 0 
        while head:
            count +=1 
            head = head.next 
        return count 
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        c = self.getLength(head)
        if n == c:
            return head.next 
        k = c - n - 1
        curr = head
        while k >  0:
            curr = curr.next 
            k-= 1
        if curr.next:
            curr.next = curr.next.next 
        return head

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int length(ListNode* head)
    {
        int count = 0;
        while(head)
        {
            count += 1;
            head = head->next;
        }
        return count;
    }
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        int k = length(head);
        k -= n;
        ListNode* dummy = new ListNode(-1);
        dummy->next = head;
        ListNode* ans = dummy;
        while(k-- && dummy)
        {
            dummy = dummy->next;
        }
        dummy->next = dummy->next->next;
        return ans->next;
    }
};